src/main/java/network/brightspots/rcv/OutputWriter.java

```diff
@@ -87,49 +89,133 @@ class ResultsWriter {
+  public static class OutputFileIdentifiers {
+    // Since sanitizing Slice IDs can cause filename collisions, ensure each non-sanitized
+    // Slice ID is given a unique sanitized name.
+    private static final Dictionary<String, String> sliceIdToUniqueSanitizedId = new Hashtable<>();
+    // This is a set of the values in sanitizerCollisionResolution to ensure there are no collisions
+    // in the resulting sanitized names.
+    private static final HashSet<String> uniqueSanitizedIds = new HashSet<>();
+    private final OutputType outputType;
+    private final boolean isSlice;
+    private final TabulateBySlice slice;
+    private final String sliceId;
+
+    OutputFileIdentifiers(OutputType outputType) {
+      this.outputType = outputType;
+      this.isSlice = false;
+      slice = null;
+      sliceId = null;
+    }
+
+    OutputFileIdentifiers(OutputType outputType, TabulateBySlice slice, String sliceId) {
+      this.outputType = outputType;
+      this.isSlice = true;
+      this.slice = slice;
+      this.sliceId = sliceId;
+    }
+
+    // getPath helper without the modifier argument
+    public Path getPath(String directory, String prefix, Integer sequentialId) {
+      return getPath(directory, prefix, null, sequentialId);
+    }
+
+    // The filename is an underscore-separated string containing all components needed
+    // to create a unique output file. The modifier is likely only used in tests, where
+    // it is set to "expected" to distinguish expected output files from actual output files.
+    public Path getPath(String directory, String prefix, String modifier, Integer sequentialId) {
+      List<String> parts = new ArrayList<>();
+      parts.add(prefix);
+      if (sequentialId != null) {
+        parts.add(sequentialId.toString());
+      }
+      if (modifier != null) {
+        parts.add(modifier);
+      }
+      if (isSlice) {
+        assert slice != null;
+        parts.add(sanitizeSliceWithoutCollisions(sliceId));
+        parts.add(slice.toLowerString());
+      }
+      parts.add(outputType.getBasename());
+
+      String filenameWithoutExt = String.join("_", parts);
+      return Path.of(directory, "%s.%s".formatted(filenameWithoutExt, outputType.getExtension()));
+    }
+
+    private String sanitizeSliceWithoutCollisions(String sliceId) {
+      String previousSanitizedSliceId = sliceIdToUniqueSanitizedId.get(sliceId);
+      String sanitizedSliceId;
+
+      if (previousSanitizedSliceId != null) {
+        sanitizedSliceId = previousSanitizedSliceId;
+      } else {
+        sanitizedSliceId = sanitizeStringForOutput(sliceId);
+
+        // In most cases, sanitizedSliceId will be unique and we can use it as-is.
+        // Here, we handle the case where it is not unique.
+        int increment = 1;
+        while (uniqueSanitizedIds.contains(sanitizedSliceId)) {
+          sanitizedSliceId = String.format("%s_%d", sanitizeStringForOutput(sliceId), increment);
+          increment++;
```

Comment:
Probably `Logger.error` here?

Reply:
I think `Logger.warning` is enough. It isn't halting. Along with this wording in the User Guide

> We sanitize precinct names in file names so that any illegal characters in the precinct name are replaced with the `_` character. If this causes name collisions (e.g. you have precinct `p1*` and precinct `p1\` , both of which are sanitized to `p_`) we append an incrementing number to ensure unique filenames (e.g. `p1_` and `p1_1`). The filename for a given precinct is not guaranteed across different tabulations. The exact precinct name for a precinct file is always explicit in the file itself.